.github/workflows/haskell-ci.yml

```diff
@@ -0,0 +1,190 @@
+# saves time and resources. For example, `cabal check` should come after `cabal
+# build`, because edits to the Cabal file are generally more rare than edits to
+# Haskell source files.
+jobs:
+  CI:
+    runs-on: ${{ inputs.os }}
+    steps:
+    - uses: actions/checkout@v4
+    - uses: haskell-actions/setup@v2
+      id: setup-haskell
+      with:
+        cabal-version: ${{ inputs.cabal }}
+        ghc-version: ${{ inputs.ghc }}
+    - name: cabal sdist
+      if: inputs.sdist
+      run: |
+        if [[ "${{ inputs.sdist }}" == true ]]; then
+          cabal sdist -o sdist
+          mkdir -p unpacked
+          tar -C unpacked -xzf sdist/*.tar.gz
+          cd unpacked/*
+        fi
+        echo "WORK_DIR=$(pwd)" >> $GITHUB_ENV
+    - uses: actions/cache/restore@v4
+      name: Restore Cabal cache
+      id: cache
+      env:
+        # Each `inputs.os` (e.g., `ubuntu-22.04-arm`) uniquely determines a
+        # `runner.arch` (e.g., ARM64), so there is no need to include the latter
+        # as part of the cache key.
+        key: cabal-${{ inputs.os }}-ghc${{ inputs.ghc }}
+      with:
+        path: |
+          ${{ steps.setup-haskell.outputs.cabal-store }}
+          ${{ env.WORK_DIR }}/dist-newstyle
+        # Use `github.ref` in the key for branch-specific caching. This is an
+        # appropriate default level of granularity. (In contrast, `github.sha`
+        # would be too granular.)
+        #
+        # The `restore-key` allows for restoring the cache from a different branch
+        # (e.g., the default branch) if none exists for the current branch (e.g.,
+        # on the first run on a new branch).
+        key: |
+          ${{ env.key }}-${{ github.ref }}
+        restore-keys: |
+          ${{ env.key }}-
+    - run: cabal configure ${{ inputs.configure-flags }} ${{ inputs.targets }}
+      working-directory: ${{ env.WORK_DIR }}
+    # Note [Parallelism]
+    # ------------------
+    #
+    # We split the build into two steps:
+    #
+    # - Build only the dependencies. Build them in parallel at the package level
+    #   (which is Cabal's default, see haskell/cabal#5776).
+    # - Build this project (a single package). Use --ghc-options=-j (in the
+    #   `configure` step) to build local packages' modules in parallel.
+    #
+    # This scheme maximizes the use of parallelism while avoiding
+    # oversubscription. See GHC proposal #540 for additional background. It would
+    # be nice to replace this with `cabal configure --semaphore`, but this only
+    # became available with GHC 9.8. To support older versions, we don't use it
+    # just yet.
+    - run: cabal build --only-dependencies ${{ inputs.targets }}
+      working-directory: ${{ env.WORK_DIR }}
+    - run: cabal build ${{ inputs.targets }}
+      working-directory: ${{ env.WORK_DIR }}
+    - run: cabal test ${{ inputs.targets }}
+      working-directory: ${{ env.WORK_DIR }}
```

Comment:
I don't think it makes sense to use the same set of `inputs.targets` for both `cabal build` and `cabal test`. For instance, one could imagine specifying `lib:my-lib` as a target, but `cabal test lib:my-lib` wouldn't work.